### ISSUE TITLE
no safeguards against conflicting goals/storage class settings during chunk removal, potential for data loss up to 3.0.104-wip

### ISSUE THREAD

**Issue by borkd** [moosefs/moosefs#158]
## Have you read through available documentation and open Github issues?
Yes

## Is this a BUG report, FEATURE request, or a QUESTION? Who is the indended audience?
BUG report

#### Your moosefs version and its origin (moosefs.com, packaged by distro, built from source, ...).
3.0.104-wip, built from source

#### Operating system (distribution) and kernel version.

Debian stretch

#### Hardware / network configuration, and underlying filesystems on master, chunkservers, and clients.

Server grade x86_64

#### How much data is tracked by moosefs master (order of magnitude)?

Only ~10M files

## Describe the problem you observed.

Background prep:
- defined a new storage class, named "purge"
- made a new subfolder and assigned "purge" to it
- moved files from keep3
- recursively changed class on the folder from keep3 to purge
- Oddly the resources tab shows right amount of files and dirs, but not a single chunk is accounted for in purge.

48hrs+ passed since then.

```
#  mfsxchgsclass -r keep3 purge .prune
.prune:
 inodes with storage class changed:              0
 inodes with storage class not changed:         17
 inodes with permission denied:                  0

# du -sxc .prune
12045820	.prune
12045820	total

# mfsgetgoal -r .prune
.prune:
 files with storage class       'purge' :         16
 directories with storage class 'purge' :          1

# mfsgetsclass -r .prune
.prune:
 files with storage class       'purge' :         16
 directories with storage class 'purge' :          1

# mfsscadmin list -l
[...]
keep3 : 3 ; admin_only: NO ; create_mode: STRICT ; create_labels: [V] , [P] , [P] ; keep_labels: [P] , [P] , [P] ; arch_labels: [P] , [P] , [P] ; arch_delay: 1d
[...]
purge : 1 ; admin_only: NO ; create_mode: LOOSE ; create_labels: [P] ; keep_labels: [P] ; arch_labels: [P] ; arch_delay: 1d
```

Mandatory screenshot:
<img width="1295" alt="gh146_class_change_issue" src="https://user-images.githubusercontent.com/12971584/49515488-fffd2b80-f864-11e8-8c1a-9cf0573d5514.png">

Why are the chunk numbers 0 for both KEEP and ARCHIVE sets in class `purge`, when this is clearly not the case? In certain automation scenarios, this could lead to unexpected behavior at best.

**Comment by borkd** [comment]
After looking at this again, I am pretty certain the root cause of this issue lies in the undocumented, and undefined behavior when goal/class changes are introduced in presence of snapshots.

I find the issue lies with the `mfsxchgsclass`, and/or `mfsgetgoal` succeeding instead of returning an error code and an explanation:
- why class change cannot be performed in a reliable manner
- and, where to look for potential cause (folders, files, chunks, whatever can be tracked to an existing snapshot). With such information in hand the admin/operator could perform expected goal change on the entire set of files AND snapshots.

Snapshots are essential feature of MooseFS, so are replication goals / class definitions. Neither is going away, so predictable behavior is paramount.

**Comment by borkd** [comment]
v3 is susceptible, have not testes on v4 yet.

**Comment by acid-maker** [comment]
Likely what you observe is effect of our simple algorithm that chooses storage class for chunk that belongs to more than one file (more than one storage class to be exact).

This behaviour is very predictable - and rather safe - the only problem is that I'm not sure that it is documented and I'm not sure that it is a good solution.

When chunk belongs to more than one storage class then for each class I calculate number of copies (ignoring labels) then I take maximum and set storage class as a simple goal equal to this maximum. In such approach number of copies is equal to maximal numer of copies defined in all storage classes for given chunk which is rather safe - but usually not exactly what user had in mind.

We are aware of this. To create better solution we need to find other (better) way to pick storage class for different storage classes (sometimes contradicting each other). Simple solution is to introduce storage class priority - in such approach we will always use class with highest priority. In case of more than one classes with the same highest priority we can choose class with highest class ID or something like this.

**Comment by borkd** [comment]
@acid-maker - While I first reported the issue I believed it was just a matter of clearly spelling out how MooseFS will do the right thing when it is asked to, **and allows admins to assign potentially conflicting class definitions to snapshots and thus the underlying chunks**, all without any feedback or warning from the CLI or CGI. 

I understand the current implementation may be simplistic and far from ideal, but changing would not be top priority *if the right thing happens every time and in a safe way*. Thus the original issue was labeled as "documentation" related to "data safety".

My classification of the problem changed drastically 2 days ago after an otherwise trivial hardware fault which completely took out a single chunkserver in the middle of space rebalancing. I would like to be wrong, but either the the algorithm is not working per your description, or there are 2nd order effects and fragility that needs to be accounted for during rebalancing.

To replay:

1. Except for reshuffling/rebalancing of chunks the MooseFS instance with master 3.0.104-wip was otherwise in a healthy steady state, with no undergoal chunks and no storage marked for removal. Each physical machine was responsible for a single chunkserver process.

2. During the time leading to and after that single chunkserver failure there were no other hardware errors anywhere within the infrastructure that I am aware of.

3. A single chunkserver in a cluster of 11 died during space rebalancing. That server hosted ~14TB worth of chunks on XFS, had ECC RAM, and ~1Gbyte/s aggregate throughput to disks. Over 3+ years I used MooseFS this, or even a simultaneous double server failure would be a non-event.

4. As expected, replication of 14TB of undergoal chunks kicked off shortly after the failure.

5. Two namespace check loops after the failure the master declared 1881 chunks as MIA and 510 files as either missing or affected. Each missing chunk was reported as "NO COPY", nothing was declared as "INVALID".

6.  "Interestingly", according to the CGI interface all of the 1881 chunks reported as unaccounted for were from RL=3. Now, **there is no other way to put it: 3 replicas (or MAX to match your description) should never, ever yield 0 copies with a fault of a single chunkserver.** Even if the current implementation is broken and leaves MIN of the intended goals, I should be left with a single chunk somewhere, and great care should be taken to make the chunk replication reasonably fail-safe. As in replicate & FSYNC (I did use `HDD_FSYNC_BEFORE_CLOSE = 1`), then after some time **lazily read/verify recently written chunks** and only then OK removal of excess copies (#10).

7. While I prefer cattle approach in my server and storage deployments, just because the machine was not a thousand miles away and it was the first time MooseFS lost data I went through the trouble of extracting the disks from the chassis and scanning the filesystems for chunks reported as missing. I was able to find 1874 chunks. and while damage was now narrowed down to 7 missing chunks tha leaves me with 7 sets of (file, snapshot) pairs unusable/corrupted.

8. I scanned the metalogger logs (~3 days worth of namespace changes) for any traces of user activity to the 14 affected files and found none.

9. Out of curiosity I kicked off a search for the filenames of missing chunks on the raw block devices. Since XFS is not a COW filesystem, there is a good chance that space used by those chunks, if they did exist on that machine, was overwritten during rebalancing (internal between disks, or from elsewhere).

Finally, I don't want to compare apples and oranges, but 3 years ago I have been bitten by a similar bug while running a rewrite of forked MooseFS that shall not be named. https://github.com/lizardfs/lizardfs/issues/298

Something can certainly be learned from all this.
To be continued..

**Comment by borkd** [comment]
So, in the process of restoring the filesystem to normal operational state and expanding capacity I noticed that fix for #146 is not complete, and that subsequent replication/rebalancing can misplace the chunks despite policy set to STRICT. In some scenarios this could lead to data loss.


A single undergoal chunk in the info tab was stuck in the (desired RL)-2 state, despite millions of other chunks being restored from RL-1 to RL. Before dismissing it as a side effect of #165 I went looking for file(s) that referenced it. In the process I noticed another single chunk file which was squatting on the special purpose chunkserver (used exclusively for shortlived CREATE mode and for CREATE and KEEP of an experimental storage class. Something that should have been fixed by 104-wip, see #146).


Search within metadata found the original file and a snapshot referencing that single chunk. Original file created on Jan 2nd, and snapshot taken on 21st:
```
0 client:/mnt/mfs/backend/backups/filemeta# ls -latr /mnt/mfs/.snaps/fs-20190121/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591 /mnt/mfs/fs/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591
-rw-r--r-- 1 root root 838 Jan  2 06:46 /mnt/mfs/fs/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591
-rw-r--r-- 1 root root 838 Jan 21 10:40 /mnt/mfs/.snaps/fs-20190121/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591
```

Storage classes are deliberately different between original file and a snapshot:
```
0 client:/mnt/mfs/backend/backups/filemeta# mfsgetsclass /mnt/mfs/.snaps/fs-20190121/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591 /mnt/mfs/fs/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591
/mnt/mfs/.snaps/fs-20190121/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591: 'keep3'
/mnt/mfs/fs/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591: 'vwip'
```

In this instance, only after chunks age to archival threshold there is some overlap between storage class definitions: 

```
0 client:/mnt/mfs/backend/backups/filemeta# mfsscadmin list -l | egrep 'keep3|vwip' 
keep3 : 2->3->4 ; admin_only: NO ; create_mode: STRICT ; create_labels: [PSZ] , [P] ; keep_labels: [PSZ] , [P] , [P] ; arch_labels: [P] , [P] , [P] , [T] ; arch_delay: 1d
vwip : 1->1->2 ; admin_only: NO ; create_mode: STRICT ; create_labels: [V] ; keep_labels: [V] ; arch_labels: [P] , [P] ; arch_delay: 1d
```

Today is 29th. Both files are old enough and should be archived. They are indeed marked as such:
```
0 client:/mnt/mfs/backend/backups/filemeta# mfschkarchive /mnt/mfs/.snaps/fs-20190121/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591 /mnt/mfs/fs/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591
/mnt/mfs/.snaps/fs-20190121/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591: all chunks are archived
/mnt/mfs/fs/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591: all chunks are archived
```

If MooseFS is doing the right thing I expect to have 4 replicas of the chunk in question, 3 on servers with label P, and 1 on server with label T.
```
0 client:/mnt/mfs/backend/backups/filemeta# mfsfileinfo /mnt/mfs/.snaps/fs-20190121/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591 /mnt/mfs/fs/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591
/mnt/mfs/.snaps/fs-20190121/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591:
        chunk 0: 000000000063CA1D_00000001 / (id:6539805 ver:1)
                copy 1: 192.168.1.35:9482 (status:VALID)
                copy 2: 192.168.1.38:9422 (status:VALID)
                copy 3: 192.168.1.40:9422 (status:VALID)
                copy 4: 192.168.1.229:9422 (status:VALID)
/mnt/mfs/fs/homes/logs/wip/v1/rm.0bf552b50f7249c813af84179aeaeecd01370f13893bcc429bb71c11d867bec5.v1.1546429564.660122591:
        chunk 0: 000000000063CA1D_00000001 / (id:6539805 ver:1)
                copy 1: 192.168.1.35:9482 (status:VALID)
                copy 2: 192.168.1.38:9422 (status:VALID)
                copy 3: 192.168.1.40:9422 (status:VALID)
                copy 4: 192.168.1.229:9422 (status:VALID)
```

We do have 4 copies at the moment. BUT, notice the copy 1 resides on .1.35 port 9482. This is a special purpose chunkserver, and under normal circumstances should always have 0 chunks. So let's confirm chunkserver labels:
```
0 client: mfscli -n -SCS -H mfsmaster | egrep '1\.35|1\.38|1\.40|1\.229'
chunk servers:	192.168.1.35	9422	8	E,H,P,U,Z	3.0.101	0	maintenance_on	1583583	4311641358336	4381928718336	-	0	0	0
chunk servers:	192.168.1.35	9482	7	H,V	3.0.104	0	maintenance_off	85	63200821248	225747271680	-	0	0	0
chunk servers:	192.168.1.38	9422	4	E,G,P,S,Z	3.0.101	1	maintenance_off	409984	456072101888	459352440832	-	0	0	0
chunk servers:	192.168.1.40	9422	6	A,E,G,P,U	3.0.100	1	maintenance_off	462688	4576172836352	4608680030720	-	0	0	0
chunk servers:	192.168.1.229	9422	15	C,G,P,S,Z	3.0.104	0	maintenance_off	1587114	5222031949824	5274180386816	-	0	0	0
```

Labels don't match class definition despite STRICT policy, not good. It looks that this copy is not a leftover from CREATE, or KEEP times, but was somehow replicated to this chunkserver yesterday (pretty bad) and left there (even worse).
```
0 chunkserver-1-35:/discard/mfscs-incoming# find -name "*.mfs" -ls
     2925      2 -rw-r-----   1 mfs      mfs         73728 Jan 28 07:43 ./04/chunk_000000000063CA1D_00000001.mfs
```

**Comment by borkd** [comment]
To make it easier to troubleshoot similar problems and monitor progress during storage class 
 or even simple RL changes please add a parameter to `mfsfileinfo` that reports the `%coverage` between replication level as set in object's storage policy and actual on-disk state for all chunk copies belonging to said file at the time of the inquiry. Alternatively `mfscoverage` could be a separate utility,  moving us towards on-demand scrub of namespace and objects #117.
